File: auth-service/src/main/java/com/eirsteir/coffeewithme/authservice/domain/Account.java
package com.eirsteir.coffeewithme.authservice.domain;

import com.eirsteir.coffeewithme.commons.domain.account.AccountCreatedEvent;
import io.eventuate.tram.events.publisher.ResultWithEvents;
import java.util.Collection;
import java.util.Collections;
import javax.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
public class Account {

  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  private Long id;

  @Column(unique = true)
  private String email;

  private String name;

  private String password;

  @ManyToMany(fetch = FetchType.EAGER)
  @JoinTable(
      name = "account_roles",
      joinColumns = @JoinColumn(name = "account_id", referencedColumnName = "id"),
      inverseJoinColumns = @JoinColumn(name = "role_id", referencedColumnName = "id"))
  private Collection<Role> roles;

  public static ResultWithEvents<Account> createAccount(Account account) {
    AccountCreatedEvent accountCreatedEvent =
        new AccountCreatedEvent(account.getId(), account.getEmail(), account.getName());
    return new ResultWithEvents<>(account, Collections.singletonList(accountCreatedEvent));
  }
}

File: auth-service/src/main/kotlin/com/eirsteir/coffeewithme/authservice/domain/Account.kt
package com.eirsteir.coffeewithme.authservice.domain

import com.eirsteir.coffeewithme.commons.domain.account.AccountCreatedEvent
import io.eventuate.tram.events.publisher.ResultWithEvents
import lombok.*
import javax.persistence.*


@Entity
data class Account(
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    val id: Long = -1L,
    @Column(unique = true)
    val email: String = "",
    val name: String = "",
    val password: String = "",
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "account_roles",
        joinColumns = [JoinColumn(name = "account_id", referencedColumnName = "id")],
        inverseJoinColumns = [JoinColumn(name = "role_id", referencedColumnName = "id")]
    )
    val roles: Set<Role>? = null
) {

    companion object {
        fun createAccount(account: Account): ResultWithEvents<Account> {
            val accountCreatedEvent = AccountCreatedEvent(account.id, account.email, account.name)
            return ResultWithEvents(account, listOf(accountCreatedEvent))
        }
    }
}